spec/app/domain/authentication/authn_k8s/application_identity_spec.rb

```diff
@@ -0,0 +1,511 @@
+
+  let(:service_account_annotation) { double("ServiceAccountAnnotation") }
+  let(:pod_annotation) { double("PodAnnotation") }
+  let(:deployment_annotation) { double("DeploymentAnnotation") }
+  let(:stateful_set_annotation) { double("StatefulSetAnnotation") }
+  let(:deployment_config_annotation) { double("DeploymentConfigAnnotation") }
+
+  let(:invalid_annotation) { double("InvalidAnnotation") }
+
+  let(:good_service_id) { "MockService" }
+
+  before(:each) do
+    allow(namespace_annotation).to receive(:values)
+                                     .and_return(namespace_annotation)
+    allow(namespace_annotation).to receive(:[])
+                                     .with(:name)
+                                     .and_return("authn-k8s/namespace")
+    allow(namespace_annotation).to receive(:[])
+                                     .with(:value)
+                                     .and_return("K8sNamespace")
+
+    allow(namespace_annotation_service_id_scoped).to receive(:values)
+                                                       .and_return(namespace_annotation_service_id_scoped)
+    allow(namespace_annotation_service_id_scoped).to receive(:[])
+                                                       .with(:name)
+                                                       .and_return("authn-k8s/#{good_service_id}/namespace")
+    allow(namespace_annotation_service_id_scoped).to receive(:[])
+                                                       .with(:value)
+                                                       .and_return("K8sNamespaceServiceIdScoped")
+
+    allow(container_name_annotation).to receive(:values)
+                                          .and_return(container_name_annotation)
+    allow(container_name_annotation).to receive(:[])
+                                          .with(:name)
+                                          .and_return("authn-k8s/authentication-container-name")
+    allow(container_name_annotation).to receive(:[])
+                                          .with(:value)
+                                          .and_return("ContainerName")
+
+    allow(container_name_annotation_service_id_prefix).to receive(:values)
+                                                            .and_return(container_name_annotation_service_id_prefix)
+    allow(container_name_annotation_service_id_prefix).to receive(:[])
+                                                            .with(:name)
+                                                            .and_return("authn-k8s/#{good_service_id}/authentication-container-name")
+    allow(container_name_annotation_service_id_prefix).to receive(:[])
+                                                            .with(:value)
+                                                            .and_return("ServiceIdContainerName")
+
+    allow(container_name_annotation_kubernetes_prefix).to receive(:values)
```

Comment:
I remember we discussed this before but what are we doing here with allow() to do something? We are specifying what parameters it can accept and once it does, return the relevant data values defined in `and_return`?